### PR TITLE
Verena reimund exercise 1

### PR DESCRIPTION
**Pull request by vreimund**
code for exercise 1, no doxygen comments

### CODE REVIEW

**Comment by OwenArnold** [comment]
Hi @reimundILL I was just looking at your changes. The windows build is failing because of your recent changes: 

```
C:\Jenkins\workspace\newstarter-pull_requests-win7\exercises\verena_reimund\ex02_oo_basics\src\main.cpp(73): error C2065: 'M_PI' : undeclared identifier [C:\Jenkins\workspace\newstarter-pull_requests-win7\builds\verena_reimund\ex02_oo_basics\Shapes.vcxproj]
```

You can fix this by adding the following line to the end of your `CMakeLists.txt`. I wouldn't have expected that you would have known this. 

```
add_definitions ( -D_USE_MATH_DEFINES )
```

Ask me or @ianbush  if this doesn't work for you.

**Comment by OwenArnold** [comment]
@reimundILL Other than that your changes are starting to look good 👍

File: exercises/verena_reimund/ex02_oo_basics/src/main.cpp [hunk 1]
```diff
@@ -1,181 +1,159 @@
...
+    return s1->area()>s2->area();
 }
 
-bool compare_perimeter (const Shapes *s1, const Shapes *s2){
-   return s1->perimeter()>s2->perimeter();
+bool compare_perimeter (const shape *s1, const shape *s2){
+    return s1->perimeter()>s2->perimeter();
+}
+
+bool compare_sides (const int &s1, const int &s2){
```
Comment (OwenArnold) [inline exercises/verena_reimund/ex02_oo_basics/src/main.cpp]: Perfect!

File: exercises/verena_reimund/ex02_oo_basics/src/main.cpp [hunk 2]
```diff
@@ -1,181 +1,159 @@
...
-   std::string n_sides;
+class shape{
 public:
-   Shapes(){}
-   virtual std::string type() const {return n;}
-   virtual std::string Square() const {return n_sides;}
-   virtual double perimeter () const {return shape_perimeter;}
-   virtual double area () const {return shape_area;}
-   virtual ~Shapes(){}
+    shape(){}
```
Comment (OwenArnold) [inline exercises/verena_reimund/ex02_oo_basics/src/main.cpp]: This is exactly the way I think it should be done. 👍 

As a side-topic, when you're using c++11, I suggest using the new `override` keyword wherever possible. You can't use it for the current exercise, but you can in Mantid.

File: exercises/verena_reimund/ex02_oo_basics/src/main.cpp [hunk 3]
```diff
@@ -1,181 +1,159 @@
...
+class triangle: public shape{
+    double height;
+    double base;
 public:
-   triangle(double h, double b) : height(h), base(b) {}
-   std::string type() const {return "Triangle";}
-   std::string Square() const {return "3";}
-   double perimeter() const {return(base+2*sqrt(pow(height,2.0)+pow(base,2.0)/4));}
-   double area() const {return (height*base/2);}
+    triangle(double h, double b) : height(h), base(b) {}
```
Comment (OwenArnold) [inline exercises/verena_reimund/ex02_oo_basics/src/main.cpp]: Also good!

**Comment by vreimund** [comment]
Hi @OwenArnold thank you for so much help and hints! It interests me. I will commit the next, final changes as soon as possible.

**Comment by OwenArnold** [comment]
Hi @reimundILL . Send me a message when you're done. Thanks!

**Comment by vreimund** [comment]
Hi @OwenArnold , I was interested to use dynamic_cast in exercise 2 and it works fine now for me.

File: exercises/verena_reimund/ex02_oo_basics/src/main.cpp [hunk 4]
```diff
@@ -8,152 +8,203 @@
...
+    virtual double Area () const =0;
+    virtual ~Shape(){}
 };
 
 // type definitions
 
-typedef std::vector<shape*> collection; // std::unique_ptr
-typedef std::vector<shape*>::size_type its;
+typedef std::vector<Shape*> collection; // std::unique_ptr
+typedef std::vector<Shape*>::size_type ShapeIterator;
```
Comment (OwenArnold) [inline exercises/verena_reimund/ex02_oo_basics/src/main.cpp:25]: Slightly unusual name as you `ShapeIterator` is not really an iterator. For example, users might expect after getting a `ShapeIterator` that they might be able to call `begin` and `end` on it. Not worth changing though.

**Comment by OwenArnold** [comment]
Overall, this is good now. I'm happy to mark this exercise as complete if @ianbush is too?

**Comment by ianbush** [comment]
This looks good to me, the counting now appears to be correct. I noticed semi-colons are not treated as punctuation, but this is a problem in the list of punctuation given in the exercise!

**Comment by OwenArnold** [comment]
@reimundILL Please add a gravitar to your github account. No other action needed.

**Pull request closed without merging**